### PR TITLE
[MNT] improvements to modular CI framework - clearer naming, `pyproject` handling

### PR DESCRIPTION
**Pull request by fkiraly**
This PR makes multiple improvements to the modular CI by @yarnabrina:

* shortened and clearer names for better recognizability in lists and tables
* a change in `pyproject.toml` now triggers all elements to be run
* more diagnostic prints: python version, full environment freeze, available git branches are now printed out in every CI run

### changes in naming

Shortens and clarifies names in the new modular CI framework created by @yarnabrina, concrete suggestion along the lines of discussion https://github.com/sktime/sktime/issues/5706#issuecomment-1879671293

Proposed naming changes:
* general shortening, removal of redundant words like "run unit tests for" or non-differentiating informationn
* "with extras" -> test module; "without extras" -> test other

### CODE REVIEW

**Comment by fkiraly** [comment]
@yarnabrina, I thought a change in `pyproject.toml` would trigger all tests - does it not?

File: .github/actions/test-component/action.yml
Comment (yarnabrina) [inline .github/actions/test-component/action.yml:1]: Does it work? Just confirming, as I'm not sure if inputs are available before name is being defined.

File: .github/actions/test-component/action.yml
Comment (yarnabrina) [inline .github/actions/test-component/action.yml:2]: Same as above, does description have access to inputs?

File: pyproject.toml
Comment (yarnabrina) [inline pyproject.toml:83]: Note: it'll trigger all the validate extras jobs, but it doesn't cover `all_extras`. Only component specific dependency sets and `all_extras_pandas2` are covered.

https://github.com/sktime/sktime/blob/5ab219917e23233bef2ee17bce600f5996edaeec/.github/workflows/validate_extras.yml#L23-L33

File: .github/workflows/test_all.yml
Comment (yarnabrina) [inline .github/workflows/test_all.yml:54]: Note: this step does not use `all_extras`, rather this is the step that uses `forecasting`, `classification`, etc. optional dependencies.

File: .github/workflows/test_all.yml
Comment (yarnabrina) [inline .github/workflows/test_all.yml:56]: Same as above

File: .github/workflows/test_all.yml
Comment (yarnabrina) [inline .github/workflows/test_all.yml:95]: This is the step that uses `all_extras_pandas2`, and doesn't use dependencies only for a specific component/module.

File: .github/workflows/test_all.yml
Comment (yarnabrina) [inline .github/workflows/test_all.yml:97]: Same as above

File: .github/workflows/test_datasets.yml
Comment (yarnabrina) [inline .github/workflows/test_datasets.yml:1]: Note: it's not VRON only. If a PR modifies `sktime/base` or `sktime/datasets`, this workflow is triggered by the code quality workflow.

**Review by yarnabrina — COMMENTED** [review]
I added few comments where it seemed necessary. Other than these, I only have two general notes:

1. the word `test` on its own are used in multiple places as workflow/job/step names, which may be confusing later to identify what is being referred to
2. The word module and other are probably used to replace "components with extras" (e.g. forecasting) and "components without extras" (e.g. benchmarking) respectively. If that's the goal, I think they may have been used in the reversed order in some cases.

File: .github/workflows/test_all.yml
Comment (fkiraly) [reply .github/workflows/test_all.yml:54]: yes, I see. Apologies, I got confused by the naming, since "with extras" is very close to "all extras", semanticlly.

File: .github/actions/test-component/action.yml
Comment (fkiraly) [reply .github/actions/test-component/action.yml:1]: I think it does not

File: .github/workflows/test_datasets.yml
Comment (fkiraly) [reply .github/workflows/test_datasets.yml:1]: updated

File: pyproject.toml
Comment (fkiraly) [reply pyproject.toml:83]: I hoped a change to `pyproject.toml` would trigger everything

**Comment by yarnabrina** [comment]
Can you please simulate changes in `sktime/base`, `sktime/datasets`, at least one "dedicated" module and at least one "non-dedicated" module so that running unit testing is validated?

**Comment by fkiraly** [comment]
@yarnabrina, done.

Since we last spoke, I also widened the scope:

* a change in `pyproject.toml` triggers all the tests (except `datasets`). This should adress the coverage gap in a case where only `pyproject.toml` changes, with possible far-reaching impact on dependent estimators. Could be combined with https://github.com/sktime/sktime/pull/5727 to cut down the load to the actually impacted ones.
* added diagnostic lines to print out the python environment and git branches.

**Review by yarnabrina — CHANGES_REQUESTED** [review]
image

I'm really not sure seeing this instead is any more clearer or helpful to know what job is running what. If everything is named `test`, is it any more easier to find for you?

`pyproject.toml` change triggering everything is also theoretically fine, but this should increase number of jobs and runtime significantly. If I add a `skforecast` adapter for example, just by adding a new specification in forecasting is now going to trigger every other component as well, and they will basically do nothing as all they will be filtered out by only-test-modified-estimators logic.

But by your description, the current CI should have tested everything, even the "components-without-extras" thing. That did not happen, and that job is skipped (`other/test_module` - https://github.com/sktime/sktime/actions/runs/7508337548/job/20446474397). Is this according to your plan/expectation?

File: .github/actions/test-base/action.yml
Comment (yarnabrina) [inline .github/actions/test-base/action.yml:34]: This action is called with a python version input, and it says so in the job name. Is there a situation where this will add extra information?

File: .github/workflows/test_base.yml
Comment (yarnabrina) [inline .github/workflows/test_base.yml:22]: I did not know name can be empty, so thanks for showing. But is it a good thing? If there's no name, how shall we know which one is causing issues if a CI fails?

File: .github/actions/test-base/action.yml
Comment (yarnabrina) [inline .github/actions/test-base/action.yml:46]: Similar doubt, can you please explain what is this for? Is it not always the name of the PR branch and `main`?

**Review by yarnabrina — COMMENTED** [review]
Added few questions.

**Comment by fkiraly** [comment]
> If everything is named test, is it any more easier to find for you?

Yes, because now I can see the variable parts much more easily - 3.8, module name, OS name, etc.

File: .github/actions/test-base/action.yml
Comment (fkiraly) [reply .github/actions/test-base/action.yml:34]: * success control
* adds the information in another place, making it easiere to access there

File: .github/workflows/test_base.yml
Comment (fkiraly) [reply .github/workflows/test_base.yml:22]: based on the GUI, it shows with a red X

File: .github/actions/test-base/action.yml
Comment (fkiraly) [reply .github/actions/test-base/action.yml:46]: This is useful to diagnose issues with, e.g., differential testing, or other utilities that call `git`.

**Comment by yarnabrina** [comment]
Not sure if you missed these two comments, can you please check?

a possible caution for CI runtime

> pyproject.toml change triggering everything is also theoretically fine, but this should increase number of jobs and runtime significantly. If I add a skforecast adapter for example, just by adding a new specification in forecasting is now going to trigger every other component as well, and they will basically do nothing as all they will be filtered out by only-test-modified-estimators logic.

a possible issue in your changes

> But by your description, the current CI should have tested everything, even the "components-without-extras" thing. That did not happen, and that job is skipped (other/test_module - https://github.com/sktime/sktime/actions/runs/7508337548/job/20446474397). Is this according to your plan/expectation?

File: .github/actions/test-component/action.yml
Comment (fkiraly) [reply .github/actions/test-component/action.yml]: It did not, so I reverted

File: .github/actions/test-component/action.yml
Comment (fkiraly) [reply .github/actions/test-component/action.yml]: see above, reverted

**Comment by fkiraly** [comment]
> But by your description, the current CI should have tested everything, even the "components-without-extras" thing. That did not happen, and that job is skipped (other/test_module - https://github.com/sktime/sktime/actions/runs/7508337548/job/20446474397). Is this according to your plan/expectation?

Good spot!

No, this was not per expectation. I think the reason was a wrong output name reference, should be fixed now.

> pyproject.toml change triggering everything is also theoretically fine, but this should increase number of jobs and runtime significantly.

Yes, good point.

However, I do think an added soft dependency should trigger all tests - at least we should test the affected module, and barring any smart parsing of `pyproject.toml`, we do not know which module is affected, so the only way to satisfy this condition is running all tests.

How about:

* we remove the `test_install` (or previously `validate_components` workflow, because *now* it is actually redundant. Since a `pyproject.toml` change now triggers all component tests, this includes install too.
* we combine with https://github.com/sktime/sktime/pull/5727, which will ensure affected estimators are also tested - which should address your concern on "they will basically do nothing as all they will be filtered out by only-test-modified-estimators logic."

> If I add a skforecast adapter for example, just by adding a new specification in forecasting is now going to trigger every other component as well, and they will basically do nothing as all they will be filtered out by only-test-modified-estimators logic.

If #5727 is merged, it would trigger:

* all components other than forecasting, it basically does nothing except install - which would be tested anyway in `validate_components` (and you added that rule to have `pyproject.toml` changes trigger the install)
* in forecasting, the new estimator is tested, because `run_tests_for_class` now picks it up after #5727

**Comment by yarnabrina** [comment]
As of now, I'm not comfortable with full understanding of all the changes in these 3 PR's (#5713 + #5727 + #5785). They overlap quite a lot and have interdependencies, and I've been unable to follow what all behaviour is being modified (completely ignoring naming stuff) and where.

I've not approved this PR becuase I am not clear in my mind so far. If this indecisiveness on my part blocks the release cycle, please let me know and I'll approve to unblock.

**Comment by fkiraly** [comment]
> I've not approved this PR becuase I am not clear in my mind so far. If this indecisiveness on my part blocks the release cycle, please let me know and I'll approve to unblock.

How about we merge this one before the release, this is just a rename, and delay #5727, #5785 aftere the release?
I think the renaming will also enable other core devs to get into the discussion more easily.

**Review by yarnabrina — APPROVED** [review]
As discussed above, approved for 0.25.1 and 0.26.0 release cycle.

Current changes as of my understanding:

1. rename most of workflows/actions/jobs/steps
2. add change in `pyproject.toml` as trigger to every all testing jobs
3. verbose printing of python and git information

**Comment by fkiraly** [comment]
> Current changes as of my understanding:

Yes, apologies for being imprecise - thsi also adds the pyprojecet tigger and diagostic steps. Thaks for the approval.

**Pull request merged**
